cardano-api/src/Cardano/Api/IPC/Monad.hs

```diff
@@ -14,57 +15,136 @@ import Control.Concurrent.STM
+      , localTxSubmissionClient = Nothing
+      }
+    )
+
+  join . first mapError <$> atomically waitResult
+
+-- | Execute a local state query expression concurrently with a chain sync.
+executeQueryLocalStateWithChainSync
+  :: LocalNodeConnectInfo mode
+  -> Maybe ChainPoint
+  -> (AcquireFailure -> e)
+  -> (NodeToClientVersion -> ExceptT e (LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO) a)
+  -> IO (ChainTip, Either e a)
+executeQueryLocalStateWithChainSync connectInfo mpoint mapError f = do
+  tmvResultLocalState <- newEmptyTMVarIO
+  tmvResultChainTip <- newEmptyTMVarIO
+
+  waitResult <- pure $ (,)
+    <$> readTMVar tmvResultChainTip
+    <*> readTMVar tmvResultLocalState
 
-          , Net.Query.recvMsgFailure = \failure -> do
-              atomically $ putTMVar resultVar' (Just (Left failure))
+  connectToLocalNodeWithVersion
+    connectInfo
+    (\ntcVersion ->
+      LocalNodeClientProtocols
+      { localChainSyncClient    = LocalChainSyncClient $ chainSyncGetCurrentTip waitResult tmvResultChainTip
+      , localStateQueryClient   = Just $ setupLocalStateQueryExpr waitResult mpoint tmvResultLocalState (runExceptT (f ntcVersion))
+      , localTxSubmissionClient = Nothing
+      }
+    )
+
+  fmap (join . first mapError) <$> atomically waitResult
+
+  where
+    chainSyncGetCurrentTip
+      :: STM b
+      -> TMVar  ChainTip
+      -> ChainSyncClient (BlockInMode mode) ChainPoint ChainTip IO ()
+    chainSyncGetCurrentTip waitDone tipVar =
+      ChainSyncClient $ pure clientStIdle
+      where
+        clientStIdle :: Net.Sync.ClientStIdle (BlockInMode mode) ChainPoint ChainTip IO ()
+        clientStIdle =
+          Net.Sync.SendMsgRequestNext clientStNext (pure clientStNext)
+
+        clientStNext :: Net.Sync.ClientStNext (BlockInMode mode) ChainPoint ChainTip IO ()
+        clientStNext = Net.Sync.ClientStNext
+          { Net.Sync.recvMsgRollForward = \_block tip -> ChainSyncClient $ do
+              void . atomically $ putTMVar tipVar tip
               void $ atomically waitDone
-              pure $ Net.Query.SendMsgDone ()
+              pure $ Net.Sync.SendMsgDone ()
+          , Net.Sync.recvMsgRollBackward = \_point tip -> ChainSyncClient $ do
+              void . atomically $ putTMVar tipVar tip
+              void $ atomically waitDone
+              pure $ Net.Sync.SendMsgDone ()
           }
-      else do
-        atomically $ putTMVar resultVar' Nothing
+
+-- | Use 'sendMsgQuery' in a do block to construct monadic local state queries.
+setupLocalStateQueryExpr ::
+     STM x
+  -> Maybe ChainPoint
+  -> TMVar (Either Net.Query.AcquireFailure a)
+  -> LocalStateQueryExpr (BlockInMode mode) ChainPoint (QueryInMode mode) () IO a
+  -> Net.Query.LocalStateQueryClient (BlockInMode mode) ChainPoint (QueryInMode mode) IO ()
+setupLocalStateQueryExpr waitDone mPointVar' resultVar' f =
```

Comment:
`waitDone` is need still because it's what keeps the protocols alive.  `waitDone` only returns when all the component `TVar`s are populated, indicating that it is okay to exit.

Specifically, it is needed by `executeQueryLocalStateWithChainSync`, but not `executeQueryLocalState`, but they share the `setupLocalStateQueryExpr` implementation, so it is retained in both cases.

Reply:
Why do we need to keep the protocols alive? Aren't the mini protocols allowed to exit at different times? I would assume that we don't need to keep the local-state-query protocol going in order to still use the chain-sync protocol.

Reply:
I believe if you get to the `Done` state, all miniprotocols terminate.  This was causing an STM deadlock previously.

Reply:
> Why do we need to keep the protocols alive? Aren't the mini protocols allowed to exit at different times? 

This indeed is a consequence of old mux api, we haven't yet updated the `ouroboros-network` client api for its new interface which would allow that.  We have an issue for this here: https://github.com/input-output-hk/ouroboros-network/issues/3231